vault/expiration.go

```diff
@@ -2291,6 +2345,47 @@ func (m *ExpirationManager) walkLeases(walkFn leaseWalkFunction) error {
 	return nil
 }
 
+// must be called with m.pendingLock held
+func (m *ExpirationManager) removeFromPending(ctx context.Context, leaseID string) {
+	if info, ok := m.pending.Load(leaseID); ok {
+		pending := info.(pendingInfo)
+		pending.timer.Stop()
+		m.pending.Delete(leaseID)
+		m.leaseCount--
+		// Log but do not fail; unit tests (and maybe Tidy on production systems)
+		if err := m.core.quotasHandleLeases(ctx, quotas.LeaseActionDeleted, []string{leaseID}); err != nil {
+			m.logger.Error("failed to update quota on revocation", "error", err)
+		}
+	}
+}
+
+// note: must be called with pending lock held
+func (m *ExpirationManager) markLeaseAsZombie(ctx context.Context, le *leaseEntry, err error) {
+	if le == nil {
+		m.logger.Warn("attempted to mark nil lease as zombie")
+		return
+	}
+	if le.isZombie() {
+		m.logger.Info("attempted to re-mark lease as zombie", "original_error", le.RevokeErr, "new_error", err.Error())
+		return
+	}
+
+	errStr := err.Error()
+	if len(errStr) == 0 {
+		errStr = "no error message given"
+	}
+	if len(errStr) > maxZombieErrorLength {
+		errStr = errStr[:maxZombieErrorLength]
+	}
+
+	le.RevokeErr = errStr
+	m.persistEntry(ctx, le)
+
+	m.zombies.Store(le.LeaseID, le)
+	m.removeFromPending(ctx, le.LeaseID)
```

Comment:
I don't understand quotas enough to comment on correctness here, but this does seem t one adding in an extra call to `quotasHandleLeases`. Could you explain why you're adding that call in? Thanks!

Reply:
But shouldn't zombie leases count to towards the lease count quota? Zombie leases seem exactly like the thing that quotas were guarding against -- namely, vault performance being degraded due to rogue applications "generating an unbounded number of leases and/or loading too much data into Vault leading to exhausting Consul storage backend's available memory" (from https://learn.hashicorp.com/tutorials/vault/resource-quotas).

Reply:
Like, if this list of zombies is growing at an alarming rate, I'd want vault to start reporting issues based on lease quotas, rather than silently falling into a degraded state.

Reply:
this feels a bit grey, but you might be right... i'll have to think on that

either way, i'd like to address it in a more related PR (this section of code was from the zombie leases PR) - there may be more work than just skipping the quota update

Reply:
is it alright with you if I create a Jira ticket, move forward with this PR without a fix, and then address the possible fix right after?

Reply:
https://hashicorp.atlassian.net/browse/VAULT-2303